File: Sources/Core/ProfileController.swift
import Foundation

/// Zeitquelle des Controllers. In Tests injiziert, damit die Abnahmekriterien
/// aus SPEC §13 ohne echtes Warten prüfbar sind.
protocol TimeSource: Sendable {
    func now() -> Date
}

struct SystemTimeSource: TimeSource {
    func now() -> Date { Date() }
}

/// Ergebnis eines manuellen Schaltvorgangs.
struct ManualSendOutcome: Equatable, Sendable {
    /// Konnte der Befehl abgesetzt werden? Eine Bestätigung, dass die Anlage
    /// ihn umgesetzt hat, gibt es prinzipbedingt nicht.
    var delivered: Bool
    /// Im Modus `sticky` das neue Grundprofil. Der Aufrufer muss es
    /// persistieren — der Controller schreibt selbst keine Einstellungen.
    var newBaseProfile: String?
}

/// Zustandsautomat aus SPEC §8 — das Herzstück.
///
/// Er kennt genau eine relevante Größe: `lastSentProfile`, das Profil, das die
/// App zuletzt selbst an das Dashboard geschickt hat. Was in der Anlage
/// tatsächlich aktiv ist, kann niemand auslesen.
///
/// Die Rückschalt-Verzögerung ist bewusst kein Timer, sondern ein Vergleich von
/// Zeitstempeln beim jeweils nächsten Poll. Das macht das Verhalten ohne echtes
/// Warten testbar und entspricht der Realität: der begrenzende Faktor ist
/// ohnehin das Poll-Intervall.
///
/// Bewusst frei von AppKit und SwiftUI.
actor ProfileController {
    private let bridge: any ProfileActivating
    private let time: any TimeSource
    /// Zuletzt gesendetes Profil und Grundprofil liegen hier, damit das
    /// Sicherheitsnetz sie beim Einschlafen synchron lesen kann.
    private let safety: SafetyNet

    private(set) var lastSentAt: Date?
    /// Letzter Sendeversuch fehlgeschlagen — speist den Warnzustand des Icons.
    private(set) var lastSendFailed = false

    var lastSentProfile: String? { safety.lastSentProfile }
    var baseProfile: String { safety.baseProfile }

    private var manualMode: ManualMode
    private var automationEnabled: Bool
    private var rules: [Rule]
    private var resetDelay: TimeInterval
    private var blindTimeout: TimeInterval

    /// Beginn der laufenden Blindphase, solange der Status unbekannt ist.
    private(set) var blindSince: Date?
    /// Seit wann steht das Zielprofil durchgehend auf dem Grundprofil?
    private(set) var resetPendingSince: Date?

    /// Zuletzt bekannte Graph-Activity; `nil` heißt, Teams ist offline.
    private var lastKnownActivity: String?
    /// Lokal erkannt, unabhängig von Teams.
    private(set) var awayFromDesk = false

    init(
        bridge: any ProfileActivating,
        time: any TimeSource = SystemTimeSource(),
        settings: Settings,
        safetyNet: SafetyNet? = nil
    ) {
        self.bridge = bridge
        self.time = time
        safety = safetyNet ?? SafetyNet(baseProfile: settings.baseProfile)
        safety.setAwayProfile(settings.awayProfile)
        manualMode = settings.manualMode
        automationEnabled = settings.automationEnabled
        rules = settings.rules
        resetDelay = TimeInterval(settings.resetDelaySeconds)
        blindTimeout = TimeInterval(settings.blindTimeoutSeconds)
    }

    /// Übernimmt geänderte Einstellungen aus dem UI.
    func apply(_ settings: Settings) {
        safety.setBaseProfile(settings.baseProfile)
        safety.setAwayProfile(settings.awayProfile)
        manualMode = settings.manualMode
        automationEnabled = settings.automationEnabled
        rules = settings.rules
        resetDelay = TimeInterval(settings.resetDelaySeconds)
        blindTimeout = TimeInterval(settings.blindTimeoutSeconds)
    }

    /// Steht gerade ein Regelprofil? Steuert das kürzere Poll-Intervall.
    var isOnRuleProfile: Bool {
        guard let lastSentProfile else { return false }
        return lastSentProfile != baseProfile
    }

    // MARK: Automatik

    func handle(_ result: PresenceResult) async {
        // Pausierte Automatik heißt: keinerlei automatisches Senden. Manuelles
        // Schalten aus dem Menü bleibt möglich.
        guard automationEnabled else { return }

        switch result {
        case .presence(_, let activity):
            blindSince = nil
            lastKnownActivity = activity
            await evaluate(reason: "Activity \(activity)")

        case .offline:
            // Teams aus bedeutet real, dass nicht telefoniert wird. Lokale
            // Auslöser können trotzdem greifen.
            blindSince = nil
            lastKnownActivity = nil
            await evaluate(reason: "Teams offline")

        case .unknown(let failure):
            await stayBlind(failure)
        }
    }

    /// Zweiter Eingang neben der Teams-Präsenz: lokal erkannte Abwesenheit.
    ///
    /// Wirkt sofort, nicht erst beim nächsten Poll — eine Bildschirmsperre soll
    /// nicht fünf Sekunden nachhängen.
    func setDeskPresence(_ presence: DeskPresence) async {
        guard awayFromDesk != presence.isAway else { return }
        awayFromDesk = presence.isAway

        switch presence {
        case .away(let reason):
            Log.info(.controller, "Nicht am Platz (\(reason.text))")
        case .atDesk:
            Log.info(.controller, "Wieder am Platz")
        }

        // Während einer Blindphase wird auch hier nicht geschaltet: ob gerade
        // ein Gespräch läuft, ist dann unbekannt — und das entscheidet, welche
        // Regel gewinnt. Der Zustand ist gemerkt und greift beim nächsten
        // bekannten Poll.
        guard automationEnabled, blindSince == nil else { return }
        await evaluate(reason: presence.isAway ? "nicht am Platz" : "wieder am Platz")
    }

    private func evaluate(reason: String) async {
        let engine = RuleEngine(rules: rules, baseProfile: baseProfile)
        let target = engine.targetProfile(
            activity: lastKnownActivity, awayFromDesk: awayFromDesk)
        await aim(at: target, reason: reason)
    }

    private func aim(at target: String, reason: String) async {
        guard target == baseProfile else {
            // Ein Regelprofil greift: sofort schalten, ohne Verzögerung.
            resetPendingSince = nil
            guard lastSentProfile != target else { return }
            await send(target, reason: reason)
            return
        }

        // Zurück auf das Grundprofil, aber erst, wenn der Zustand durchgehend
        // angehalten hat. Kommt zwischendurch wieder ein Treffer, wird der
        // angefangene Zeitraum oben verworfen.
        //
        // Hat die App noch gar nichts gesendet, bleibt sie still: sie weiß dann
        // nicht, was in der Anlage steht, und hat auch nichts zu korrigieren.
        // Ein von Hand am Telefon gesetztes Profil überlebt so einen App-Start.
        guard let lastSent = lastSentProfile, lastSent != baseProfile else {
            resetPendingSince = nil
            return
        }

        let now = time.now()
        guard let since = resetPendingSince else {
            resetPendingSince = now
            return
        }
        guard now.timeIntervalSince(since) >= resetDelay else { return }

        await send(baseProfile, reason: "\(reason), Verzögerung abgelaufen")
        resetPendingSince = nil
    }

    /// Bei unbekanntem Status wird **nicht** geschaltet. Das ist die wichtigste
    /// Regel der ganzen App.
    private func stayBlind(_ failure: PollFailure) async {
        // Ein angefangener Rückschalt-Zeitraum zählt nicht weiter: sonst würde
        // nach einem Netzausfall blind geschaltet.
        resetPendingSince = nil

        let now = time.now()
        guard let since = blindSince else {
            blindSince = now
            return
        }
        guard now.timeIntervalSince(since) >= blindTimeout else { return }
        // Steht bereits das Grundprofil, gibt es nichts zu retten.
        guard let last = lastSentProfile, last != baseProfile else { return }

        let minutes = Int(now.timeIntervalSince(since) / 60)
        Log.notice(.controller, """
            Sicherheitsrückfall: Status seit \(minutes) min unbekannt (\(failure.text)), \
            zurück auf "\(baseProfile)" — sonst bliebe das Telefon umgeleitet
            """)
        await send(baseProfile, reason: "Sicherheitsrückfall nach Blindphase")
    }

    /// Ende der Arbeitszeit: einmal aufräumen, danach schweigen.
    ///
    /// Was die App selbst verstellt hat, wird zurückgenommen — sonst bliebe das
    /// Telefon über Nacht umgeleitet. Anschließend werden die Signalquellen
    /// abgeschaltet, es kommt also ohnehin nichts mehr an.
    func standDown() async {
        resetPendingSince = nil
        blindSince = nil
        lastKnownActivity = nil
        awayFromDesk = false

        guard let lastSent = lastSentProfile, lastSent != baseProfile else { return }
        await send(baseProfile, reason: "Ende der Arbeitszeit")
    }

    // MARK: Manuelles Schalten

    @discardableResult
    func sendManual(profile: String) async -> ManualSendOutcome {
        let delivered = await send(profile, reason: "manuell, Modus \(manualMode.rawValue)")
        var outcome = ManualSendOutcome(delivered: delivered, newBaseProfile: nil)

        // Nach einer manuellen Auswahl darf kein alter Rückschalt-Zeitraum
        // nachwirken.
        resetPendingSince = nil

        guard delivered, manualMode == .sticky, profile != baseProfile else { return outcome }
        safety.setBaseProfile(profile)
        outcome.newBaseProfile = profile
        Log.info(.controller, "Grundprofil wandert auf \"\(profile)\" (Modus sticky)")
        return outcome
    }

    /// Testen-Knopf aus den Einstellungen. Schaltet echt — anders wäre der Test
    /// wertlos — und zählt deshalb auch als zuletzt gesendetes Profil.
    @discardableResult
    func sendTest(profile: String) async -> Bool {
        await send(profile, reason: "Testen-Knopf")
    }

    // MARK: Intern

    @discardableResult
    private func send(_ profile: String, reason: String) async -> Bool {
        Log.info(.controller, "Sende \"\(profile)\" (\(reason))")
        let delivered = await bridge.activate(profile: profile)
        lastSendFailed = !delivered
        guard delivered else { return false }
        safety.recordSent(profile)
        lastSentAt = time.now()
        return true
    }
}

File: windows/src/AgfeoPresenceBridge.Core/ProfileController.cs
namespace AgfeoPresenceBridge.Core;

/// <summary>Schaltet ein Rufprofil. Eine Bestätigung liefert die Anlage nicht.</summary>
public interface IProfileActivator
{
    Task<bool> ActivateAsync(string profileName);
}

/// <summary>Zeitquelle — in Tests gestellt, damit Zeitverhalten ohne Warten prüfbar ist.</summary>
public interface IClock
{
    DateTime Now { get; }
}

public sealed class SystemClock : IClock
{
    public DateTime Now => DateTime.Now;
}

public sealed record ManualSendOutcome(bool Delivered, string? NewBaseProfile);

public sealed record SwitchRecord(string Profile, string Reason, DateTime At, bool Delivered);

/// <summary>
/// Zustandsautomat — das Herzstück.
/// </summary>
/// <remarks>
/// Er kennt genau eine relevante Größe: das zuletzt selbst gesendete Profil.
/// Was in der Anlage tatsächlich aktiv ist, kann niemand auslesen.
///
/// Die Rückschalt-Verzögerung ist bewusst kein Timer, sondern ein Vergleich von
/// Zeitstempeln beim jeweils nächsten Poll. Das macht das Verhalten ohne echtes
/// Warten testbar und entspricht der Realität: Der begrenzende Faktor ist
/// ohnehin das Poll-Intervall.
///
/// Alle Zustandsänderungen laufen über ein Schloss, weil Poller, Anrufereignisse
/// und Oberfläche aus verschiedenen Threads hereinkommen.
/// </remarks>
public sealed class ProfileController(
    IProfileActivator bridge,
    Settings settings,
    IClock? clock = null)
{
    private readonly IClock _clock = clock ?? new SystemClock();
    private readonly SemaphoreSlim _gate = new(1, 1);

    private Settings _settings = settings;
    private string? _lastKnownActivity;
    private DateTime? _blindSince;
    private DateTime? _resetPendingSince;

    public string? LastSentProfile { get; private set; }
    public DateTime? LastSentAt { get; private set; }
    public bool LastSendFailed { get; private set; }
    public bool AwayFromDesk { get; private set; }
    public string? HeldProfile { get; private set; }
    public string BaseProfile { get; private set; } = settings.BaseProfile;

    private readonly List<SwitchRecord> _history = [];
    public IReadOnlyList<SwitchRecord> History
    {
        get { lock (_history) return _history.ToArray(); }
    }

    /// <summary>Steht gerade ein Regelprofil? Steuert das kürzere Poll-Intervall.</summary>
    public bool IsOnRuleProfile => LastSentProfile is not null && LastSentProfile != BaseProfile;

    /// <summary>
    /// Das Profil, das vor dem Verschwinden noch gesendet werden muss — oder
    /// <c>null</c>, wenn ohnehin das Grundprofil steht.
    /// </summary>
    /// <remarks>
    /// Synchron abfragbar, weil beim Herunterfahren keine Zeit für Umwege
    /// bleibt. Was das Programm nicht selbst verstellt hat, fasst es auch nicht
    /// an.
    /// </remarks>
    public string? ProfileNeededOnExit =>
        LastSentProfile is not null && LastSentProfile != BaseProfile ? BaseProfile : null;

    public async Task ApplyAsync(Settings updated)
    {
        await _gate.WaitAsync();
        try
        {
            _settings = updated;
            BaseProfile = updated.BaseProfile;
        }
        finally { _gate.Release(); }
    }

    // MARK: Automatik

    public async Task HandleAsync(PresenceResult result)
    {
        await _gate.WaitAsync();
        try
        {
            // Pausierte Automatik heißt: kein automatisches Senden. Manuelles
            // Schalten bleibt möglich.
            if (!_settings.AutomationEnabled) return;

            switch (result)
            {
                case PresenceResult.Known known:
                    _blindSince = null;
                    _lastKnownActivity = known.Activity;
                    await EvaluateAsync($"Activity {known.Activity}");
                    break;

                case PresenceResult.Offline:
                    // Teams aus bedeutet real, dass nicht telefoniert wird.
                    // Lokale Auslöser können trotzdem greifen.
                    _blindSince = null;
                    _lastKnownActivity = null;
                    await EvaluateAsync("Teams offline");
                    break;

                case PresenceResult.Unknown unknown:
                    await StayBlindAsync(unknown.Failure);
                    break;
            }
        }
        finally { _gate.Release(); }
    }

    /// <summary>
    /// Zweiter Eingang neben der Teams-Präsenz: lokal erkannte Abwesenheit.
    /// Wirkt sofort, nicht erst beim nächsten Poll.
    /// </summary>
    public async Task SetDeskPresenceAsync(DeskPresence presence)
    {
        await _gate.WaitAsync();
        try
        {
            if (AwayFromDesk == presence.IsAway) return;
            AwayFromDesk = presence.IsAway;

            Log.Info(presence is DeskPresence.Away away
                ? $"Nicht am Platz ({away.Reason.Text()})"
                : "Wieder am Platz");

            // Während einer Blindphase wird auch hier nicht geschaltet: Ob
            // gerade ein Gespräch läuft, ist dann unbekannt — und das
            // entscheidet, welche Regel gewinnt. Der Zustand ist gemerkt und
            // greift beim nächsten bekannten Poll.
            if (!_settings.AutomationEnabled || _blindSince is not null) return;
            await EvaluateAsync(presence.IsAway ? "nicht am Platz" : "wieder am Platz");
        }
        finally { _gate.Release(); }
    }

    private async Task EvaluateAsync(string reason)
    {
        var engine = new RuleEngine(_settings.Rules, BaseProfile);
        await AimAsync(engine.TargetProfile(_lastKnownActivity, AwayFromDesk), reason);
    }

    private async Task AimAsync(string target, string reason)
    {
        // Ein befristet gehaltenes Profil hat Vorrang vor allem Automatischen.
        if (HeldProfile is not null) return;

        if (target != BaseProfile)
        {
            // Ein Regelprofil greift: sofort schalten, ohne Verzögerung.
            _resetPendingSince = null;
            if (LastSentProfile == target) return;
            await SendAsync(target, reason);
            return;
        }

        // Zurück auf das Grundprofil, aber erst, wenn der Zustand durchgehend
        // angehalten hat. Hat die App noch gar nichts gesendet, bleibt sie
        // still: Sie weiß dann nicht, was in der Anlage steht, und hat auch
        // nichts zu korrigieren.
        if (LastSentProfile is null || LastSentProfile == BaseProfile)
        {
            _resetPendingSince = null;
            return;
        }

        DateTime now = _clock.Now;
        if (_resetPendingSince is null) { _resetPendingSince = now; return; }
        if ((now - _resetPendingSince.Value).TotalSeconds < _settings.ResetDelaySeconds) return;

        await SendAsync(BaseProfile, $"{reason}, Verzögerung abgelaufen");
        _resetPendingSince = null;
    }

    /// <summary>
    /// Bei unbekanntem Status wird <b>nicht</b> geschaltet. Das ist die
    /// wichtigste Regel der ganzen Anwendung.
    /// </summary>
    private async Task StayBlindAsync(PollFailure failure)
    {
        // Ein angefangener Rückschalt-Zeitraum zählt nicht weiter: Sonst würde
        // nach einem Netzausfall blind geschaltet.
        _resetPendingSince = null;

        DateTime now = _clock.Now;
        if (_blindSince is null) { _blindSince = now; return; }
        if ((now - _blindSince.Value).TotalSeconds < _settings.BlindTimeoutSeconds) return;
        if (LastSentProfile is null || LastSentProfile == BaseProfile) return;

        int minutes = (int)(now - _blindSince.Value).TotalMinutes;
        Log.Notice($"Sicherheitsrückfall: Status seit {minutes} min unbekannt "
                   + $"({failure.Text}), zurück auf \"{BaseProfile}\"");
        await SendAsync(BaseProfile, "Sicherheitsrückfall nach Blindphase");
    }

    // MARK: Manuelles Schalten

    public async Task<ManualSendOutcome> SendManualAsync(string profile, bool holdsAutomation = false)
    {
        await _gate.WaitAsync();
        try
        {
            string note = holdsAutomation ? "manuell, befristet" : $"manuell, Modus {_settings.ManualMode}";
            bool delivered = await SendAsync(profile, note);

            _resetPendingSince = null;
            HeldProfile = delivered && holdsAutomation ? profile : null;

            if (!delivered || _settings.ManualMode != ManualMode.Sticky || profile == BaseProfile)
                return new ManualSendOutcome(delivered, null);

            BaseProfile = profile;
            Log.Info($"Grundprofil wandert auf \"{profile}\" (Modus sticky)");
            return new ManualSendOutcome(delivered, profile);
        }
        finally { _gate.Release(); }
    }

    /// <summary>Testen-Knopf: schaltet echt, sonst wäre der Test wertlos.</summary>
    public async Task<bool> SendTestAsync(string profile)
    {
        await _gate.WaitAsync();
        try { return await SendAsync(profile, "Testen-Knopf"); }
        finally { _gate.Release(); }
    }

    /// <summary>Ende der Befristung: die Automatik übernimmt wieder.</summary>
    public async Task ReleaseHoldAsync()
    {
        await _gate.WaitAsync();
        try
        {
            if (HeldProfile is null) return;
            HeldProfile = null;
            Log.Info("Befristung abgelaufen, Automatik übernimmt wieder");
            await EvaluateAsync("Befristung abgelaufen");
        }
        finally { _gate.Release(); }
    }

    /// <summary>
    /// Das Dashboard wurde neu gestartet und kennt unseren Stand nicht mehr.
    /// Weil es keinen Rückkanal gibt, ist erneutes Senden die einzige
    /// Möglichkeit, wieder sicher zu sein.
    /// </summary>
    public async Task ResendLastProfileAsync()
    {
        await _gate.WaitAsync();
        try
        {
            if (LastSentProfile is null) return;
            await SendAsync(LastSentProfile, "Dashboard neu gestartet");
        }
        finally { _gate.Release(); }
    }

    /// <summary>Ende der Arbeitszeit: einmal aufräumen, danach schweigen.</summary>
    public async Task StandDownAsync()
    {
        await _gate.WaitAsync();
        try
        {
            _resetPendingSince = null;
            _blindSince = null;
            _lastKnownActivity = null;
            AwayFromDesk = false;

            if (LastSentProfile is null || LastSentProfile == BaseProfile) return;
            await SendAsync(BaseProfile, "Ende der Arbeitszeit");
        }
        finally { _gate.Release(); }
    }

    private async Task<bool> SendAsync(string profile, string reason)
    {
        Log.Info($"Sende \"{profile}\" ({reason})");
        bool delivered = await bridge.ActivateAsync(profile);
        LastSendFailed = !delivered;

        lock (_history)
        {
            _history.Insert(0, new SwitchRecord(profile, reason, _clock.Now, delivered));
            if (_history.Count > 5) _history.RemoveRange(5, _history.Count - 5);
        }

        if (!delivered) return false;
        LastSentProfile = profile;
        LastSentAt = _clock.Now;
        return true;
    }
}
